packages/editor/src/hooks/align.js

```diff
@@ -43,13 +44,18 @@ export function addAttribute( settings ) {
  * Returns an array of valid alignments for a block type depending on its
  * defined supports. Returns an empty array if block does not support align.
  *
- * @param  {string}   blockName Block name to check
- * @return {string[]}           Valid alignments for block
+ * @param  {string}   blockName                   Block name to check
+ * @param  {?boolean} considerWideControlsEnabled True if the function should consider wide and full alignments as supported.
+ * @return {string[]}                             Valid alignments for block
  */
-export function getBlockValidAlignments( blockName ) {
+export function getBlockValidAlignments( blockName, considerWideControlsEnabled = true ) {
 	// Explicitly defined array set of valid alignments
 	const blockAlign = getBlockSupport( blockName, 'align' );
 	if ( Array.isArray( blockAlign ) ) {
+		// remove wide and full from the array of alignments if theme does not supports them.
+		if ( ! considerWideControlsEnabled ) {
```

Comment:
Kinda seems like we could reorganize the function to avoid considering this twice.

```js
export function getBlockValidAlignments( blockName, hasWideEnabled = true ) {
	let validAlignments;

	const blockAlign = getBlockSupport( blockName, 'align' );
	if ( Array.isArray( blockAlign ) ) {
		// Explicitly defined array set of valid alignments
		validAlignments = blockAlign;
		return blockAlign;
	} else if ( blockAlign === true ) {
		// `true` includes all alignments...
		validAlignments = [ 'left', 'center', 'right' ];

		if ( hasBlockSupport( blockName, 'alignWide', true ) ) {
			validAlignments.push( 'wide', 'full' );
		}
	} else {
		validAlignments = [];
	}

	// Remove wide and full if not supported by theme.
	if ( ! hasWideEnabled ) {
		return reject( blockAlign, ( align ) => (
			align === 'wide' ||
			align === 'full'
		) );
	}

	return validAlignments;
}
```

Reply:
In-fact, this makes me think: How do we handle the case where `align` is specified as an array, but `alignWide` is not also provided as a supports?  Would that be something like:

```js
export function getBlockValidAlignments( blockName, hasWideEnabled = true ) {
	let validAlignments;

	const blockAlign = getBlockSupport( blockName, 'align' );
	if ( Array.isArray( blockAlign ) ) {
		// Explicitly defined array set of valid alignments
		validAlignments = blockAlign;
		return blockAlign;
	} else if ( blockAlign === true ) {
		// `true` includes all alignments...
		validAlignments = [ 'left', 'center', 'right', 'wide', 'full' ];
	} else {
		validAlignments = [];
	}

	// Remove wide and full if not supported by theme.
	const hasWideBlockSupport = hasBlockSupport( blockName, 'alignWide', true );
	if ( ! hasWideEnabled || ! hasWideBlockSupport ) {
		return reject( blockAlign, ( align ) => (
			align === 'wide' ||
			align === 'full'
		) );
	}

	return validAlignments;
}
```

It might be strange to reject `wide` and `full` if the block explicitly assigns them as `align` anyways, but it becomes more a question then if we could even rely on `alignWide` supports verbatim if it's not explicitly provided by the block definition. This way it becomes more obviously problematic to the block implementer that they'd need to provide the value.

Alternatively, we could enhance `hasBlockSupport` for the `'alignWide'` key to consider the array-form of `align` and whether it includes the `'wide'` and `'full'` values.

Thoughts?

Reply:
Hi, @aduth I applied a huge refactor to the function taking into account your suggestions.
To note that, if the theme does not support wide alignments they are ignored in the hook even if they were set using an array. While the block support for wide alignments does not apply if the block set the align using an array.

I followed this rule because currently, the toolbar removes the wide align buttons if the theme does not support wide alignment (even if the align was set using an array). But if an align previously existed it was set anyway in the editor; this caused a bug:
Use a theme that supports wide aligns.
Create a post add a columns block, set full alignment.
Use a theme that does not support wide aligns.
Open the previous post, see that the columns contain full alignment and the wide align buttons are not shown so we can not disable this alignment even if the theme does not support it.

The idea of this PR is to make sure withDataAlign follows the same rules already being followed by the alignment toolbar.